src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs

```diff
@@ -67,17 +67,51 @@ internal class SqlServerSearchService : SearchService
             _logger = logger;
         }
 
-        protected override Task<SearchResult> SearchInternalAsync(SearchOptions searchOptions, CancellationToken cancellationToken)
+        protected override async Task<SearchResult> SearchInternalAsync(SearchOptions searchOptions, CancellationToken cancellationToken)
         {
-            return SearchImpl(searchOptions, false, cancellationToken);
+            using (var connection = new SqlConnection(_configuration.ConnectionString))
+            {
+                connection.Open();
+
+                SearchResult searchResult;
+
+                // If we should include the total count of matching search results
+                if (searchOptions.IncludeTotal == TotalType.Accurate && !searchOptions.CountOnly)
+                {
+                    // Begin a transaction so we can perform two atomic reads.
+                    using (var transaction = connection.BeginTransaction(IsolationLevel.ReadCommitted))
```

Comment:
Shouldn't a higher level of isolation be used here for count to be same as number of results?

Reply:
That is a good point. The database is configured with read committed snapshot isolation, so each read by default will be based on a consistent point in time. It was my impression that this was based on the start of the transaction, but it looks like it is from the start of each statement, even when running in a transaction (please confirm). So we will want, and will will need to enable, snapshot isolation. This will give the desired behavior without blocking writes. 
Apologies, @rotodd, it looks like I didn't get this completely right.

Reply:
@johnstairs I did some reading to confirm, and you're right - specifying `IsolationLevel.ReadCommitted` with the setting `READ_COMMITTED_SNAPSHOT ON` would definitely work. I noticed that specifying `IsolationLevel.Snapshot` with the setting `ALLOW_SNAPSHOT_ISOLATION ON` would also work.

I decided to go with the second option because the first one will affect all transactions where the isolation level is either unset or set to ReadCommitted. The second option seemed to be more explicit with fewer potential side effects (more info [here](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql/snapshot-isolation-in-sql-server?redirectedfrom=MSDN)).

Since John is on vacation, @brandonpollett or @deepak-ms may you take a look at 3c160b1?

Reply:
@rotodd, what i understand from sql docs is that first option won't work for this scenario since (as John also pointed) row snapshot is from start of the statement (not transaction). So only second option will work here.

Reply:
FWIW, 
i got confused by "If you set the READ_COMMITTED_SNAPSHOT database option to ON, the database engine uses row versioning and **snapshot** isolation as the default, instead of using locks to protect the data." in the docs.
So I just tested it out. 
Result : If both ALLOW_SNAPSHOT_ISOLATION and READ_COMMITTED_SNAPSHOT are ON,
then ReadCommitted still behaves as ReadCommitted (reads based on start of **statement**).

Reply:
Ah interesting. Thank you! I was mislead by that as well. I'm interested to hear about how you tested it?

Reply:
@rotodd, used 2 interleaving transactions for testing. [This](https://www.interfacett.com/blogs/understanding-isolation-levels-sql-server-2008-r2-2012-examples/) doc might be useful.